src/session.rs

```diff
@@ -218,39 +264,55 @@ impl Session {
-                    cookie_builder.expires(time::OffsetDateTime::from_unix_timestamp(expires));
-            }
-            if let Some(ref path) = c.path {
-                cookie_builder = cookie_builder.path(path);
-            }
-            if let Some(secure) = c.secure {
-                cookie_builder = cookie_builder.secure(secure);
-            }
-            cookies.push(cookie_builder.finish());
+        match &self.content.cookies {
+            Cookies::Map(_) => unreachable!(),
+            Cookies::List(cookies) => cookies
+                .iter()
+                .map(|cookie| {
+                    let mut cookie_builder =
+                        cookie_crate::Cookie::build(cookie.name.clone(), cookie.value.clone());
+                    if let Some(expires) = cookie.expires {
+                        cookie_builder = cookie_builder
+                            .expires(time::OffsetDateTime::from_unix_timestamp(expires));
+                    }
+                    if let Some(path) = &cookie.path {
+                        cookie_builder = cookie_builder.path(path.clone());
+                    }
+                    if let Some(secure) = cookie.secure {
+                        cookie_builder = cookie_builder.secure(secure);
+                    }
+                    if let Some(domain) = &cookie.domain {
+                        cookie_builder = cookie_builder.domain(domain.clone());
+                    }
+                    cookie_builder.finish()
+                })
+                .collect(),
         }
-        cookies
     }
 
     pub fn save_cookies(&mut self, cookies: Vec<cookie_crate::Cookie>) {
-        self.content.cookies.clear();
+        let session_cookies = match self.content.cookies {
+            Cookies::Map(_) => unreachable!(),
+            Cookies::List(ref mut cookies) => cookies,
+        };
+
+        session_cookies.clear();
+
         for cookie in cookies {
-            self.content.cookies.insert(
-                cookie.name().into(),
-                Cookie {
-                    value: cookie.value().into(),
-                    expires: cookie
-                        .expires()
-                        .and_then(|v| v.datetime())
-                        .map(|v| v.unix_timestamp()),
-                    path: cookie.path().map(Into::into),
-                    secure: cookie.secure(),
-                },
-            );
+            session_cookies.push(Cookie {
+                name: cookie.name().into(),
+                value: cookie.value().into(),
+                expires: cookie
+                    .expires()
+                    .and_then(|v| v.datetime())
+                    .map(|v| v.unix_timestamp()),
+                path: cookie.path().map(Into::into),
+                secure: cookie.secure(),
+                domain: cookie
+                    .domain()
+                    .map(Into::into)
+                    .or_else(|| self.host.clone()),
```

Comment:
1. What if the cookie was set by a domain different from `self.host`?
2. The Cookie domain cannot be an IP address e.g 127.0.0.1. See https://stackoverflow.com/a/30676300/5915221